packages/server/lib/browsers/memory/index.ts

```diff
@@ -0,0 +1,293 @@
+import debugModule from 'debug'
+import _ from 'lodash'
+import si from 'systeminformation'
+import os from 'os'
+import fs from 'fs-extra'
+import { performance } from 'perf_hooks'
+import { groupCyProcesses, Process } from '../../util/process_profiler'
+import pid from 'pidusage'
+
+import type { Automation } from '../../automation'
+import path from 'path'
+
+const debug = debugModule('cypress:server:browsers:memory')
+const debugVerbose = debugModule('cypress-verbose:server:browsers:memory')
+
+const MEMORY_THRESHOLD_PERCENTAGE = Number(process.env.CYPRESS_INTERNAL_MEMORY_THRESHOLD_PERCENTAGE) || 50
+const MEMORY_PROFILER_INTERVAL = Number(process.env.CYPRESS_INTERNAL_MEMORY_PROFILER_INTERVAL) || 1000
+const MEMORY_FOLDER = process.env.CYPRESS_INTERNAL_MEMORY_FOLDER_PATH || path.join('cypress', 'logs', 'memory')
+const KIBIBYTE = 1024
+const FOUR_GIBIBYTES = 4 * (KIBIBYTE ** 3)
+
+let rendererProcess: Process | null
+let handler: MemoryHandler
+let totalMemoryLimit: number
+let jsHeapSizeLimit: number
+let started = false
+let cumulativeStats: { [key: string]: any }[] = []
+let collectGarbageOnNextTest = false
+let timer: NodeJS.Timeout | null
+let currentSpecFileName: string | null
+let statsLog: { [key: string]: any } = {}
+let gcLog: { [key: string]: any } = {}
+
+export type MemoryHandler = {
+  getTotalMemoryLimit: () => Promise<number>
+  getAvailableMemory: (totalMemoryLimit: number, log?: { [key: string]: any }) => Promise<number>
+}
+
+/**
+ * Algorithm:
+ *
+ * When the test runs starts:
+ *   1. set total mem limit for the container/host by reading off cgroup memory limits (if available) otherwise use os.totalmem()
+ *
+ * On a defined interval (e.g. 1s):
+ *   1. set current mem available for the container/host by reading off cgroup memory usage (if available) otherwise use si.mem().available
+ *   2. set current renderer mem usage
+ *   3. set max avail render mem to minimum of v8 heap size limit and total available mem (current available mem + current renderer mem usage)
+ *   4. calc % of memory used, current renderer mem usage / max avail render mem
```

Comment:
Should we add a step to do an emergency GC during the test if the memory threshold exceeds a higher percent something like 95%?

Reply:
This probably requires more discussion and doesn't need to be addressed on this PR.